In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df_User_Id = pd.read_csv('/content/drive/MyDrive/Netfilx Dataset/combined_data_1.txt', header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1])
df_User_Id

,Cust_Id,Rating
0,1:,NaN
1,1488844,3.0
2,822109,5.0
3,885013,4.0
4,30878,4.0
...,...,...
24058258,2591364,2.0
24058259,1791000,2.0
24058260,512536,5.0
24058261,988963,3.0


In [ ]:
df_User_Id.isnull().sum()

,0
Cust_Id,0
Rating,4499


In [ ]:
movie_count = df_User_Id.isnull().sum()['Rating']
movie_count

4499

In [ ]:
total_movies = df_User_Id['Cust_Id'].nunique()
total_movies

475257

In [ ]:
customer_count = total_movies - movie_count
customer_count

470758

In [ ]:
ratings_count = df_User_Id.groupby('Rating')['Rating'].agg('count')
ratings_count

,Rating
Rating,
1.0,1118186
2.0,2439073
3.0,6904181
4.0,8085741
5.0,5506583


In [ ]:
movie_id = None
movie_list = []

for i in df_User_Id['Cust_Id']:
  if ':' in i:
    movie_id = int(i.replace(':',''))
  movie_list.append((movie_id))

In [ ]:
movie_list

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,


In [ ]:
df_User_Id['movie_Id'] = movie_list
df_User_Id

,Cust_Id,Rating,movie_Id
0,1:,NaN,1
1,1488844,3.0,1
2,822109,5.0,1
3,885013,4.0,1
4,30878,4.0,1
...,...,...,...
24058258,2591364,2.0,4499
24058259,1791000,2.0,4499
24058260,512536,5.0,4499
24058261,988963,3.0,4499


In [ ]:
df_User_Id = df_User_Id[df_User_Id['Rating'].notna()]

In [ ]:
df_titles = pd.read_csv('/content/drive/MyDrive/Netfilx Dataset/movies (1) (1).csv', encoding="ISO-8859-1", header=None, names=['movie_Id', 'Title', 'Genre'], usecols=[0,1,2])
df_titles

,movie_Id,Title,Genre
0,movieId,title,genres
1,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
2,2,Jumanji (1995),Adventure|Children|Fantasy
3,3,Grumpier Old Men (1995),Comedy|Romance
4,4,Waiting to Exhale (1995),Comedy|Drama|Romance
...,...,...,...
27274,131254,Kein Bund für's Leben (2007),Comedy
27275,131256,"Feuer, Eis & Dosenbier (2002)",Comedy
27276,131258,The Pirates (2014),Adventure
27277,131260,Rentun Ruusu (2001),(no genres listed)


In [ ]:
df_titles.drop(df_titles.index[0], inplace=True)
df_titles

,movie_Id,Title,Genre
1,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
2,2,Jumanji (1995),Adventure|Children|Fantasy
3,3,Grumpier Old Men (1995),Comedy|Romance
4,4,Waiting to Exhale (1995),Comedy|Drama|Romance
5,5,Father of the Bride Part II (1995),Comedy
...,...,...,...
27274,131254,Kein Bund für's Leben (2007),Comedy
27275,131256,"Feuer, Eis & Dosenbier (2002)",Comedy
27276,131258,The Pirates (2014),Adventure
27277,131260,Rentun Ruusu (2001),(no genres listed)


In [ ]:
for col in df_titles.columns:
  if col == 'movie_Id':
    df_titles[col] = pd.to_numeric(df_titles['movie_Id'], errors='coerce')

In [ ]:
df_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27278 entries, 1 to 27278
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_Id  27278 non-null  int64 
 1   Title     27278 non-null  object
 2   Genre     27278 non-null  object
dtypes: int64(1), object(2)
memory usage: 639.5+ KB


In [ ]:
df = pd.merge(df_User_Id, df_titles, on='movie_Id', how='inner')
df

,Cust_Id,Rating,movie_Id,Title,Genre
0,1488844,3.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,822109,5.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
2,885013,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
3,30878,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
4,823519,3.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
...,...,...,...,...,...
23813498,2591364,2.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813499,1791000,2.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813500,512536,5.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813501,988963,3.0,4499,Dirty Rotten Scoundrels (1988),Comedy


In [ ]:
df.duplicated().sum()

0

# List of most popular and like genre

In [ ]:
df['movie_Id'].value_counts().head(10)

,count
movie_Id,
1905,193941
2152,162597
3860,160454
4432,156183
571,154832
3938,153996
4306,151292
2452,149866
1962,145519


In [ ]:
df[(df['Rating'] >= 4) & (df['movie_Id'] == 1905)]['Genre']

,Genre
9468531,Drama
9468532,Drama
9468533,Drama
9468536,Drama
9468537,Drama
...,...
9662464,Drama
9662465,Drama
9662467,Drama
9662469,Drama


'Drama' is most popular and liked genre from above  

# Now we are filtering data by movies and genres having best ratings

In [ ]:
df_pop = df[df['Rating'] >= 4]
df_pop

,Cust_Id,Rating,movie_Id,Title,Genre
1,822109,5.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
2,885013,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
3,30878,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
6,124105,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
8,1842128,4.0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
...,...,...,...,...,...
23813489,2092745,5.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813490,555962,5.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813493,272857,4.0,4499,Dirty Rotten Scoundrels (1988),Comedy
23813497,2562830,4.0,4499,Dirty Rotten Scoundrels (1988),Comedy


In [ ]:
df_pop= pd.DataFrame(df_pop)
df_user_Summary = df_pop.groupby('Cust_Id')['Rating'].agg(['count'])
df_user_Summary

,count
Cust_Id,
10,23
1000004,1
1000027,23
1000033,43
1000035,16
...,...
999964,38
999972,28
999977,11


In [ ]:
drop_user = []

In [ ]:
user_benchmark = round(df_user_Summary['count'].quantile(0.75),0)
user_benchmark

38.0

In [ ]:
drop_user = df_user_Summary[df_user_Summary['count'] < user_benchmark].index
drop_user

Index(['10', '1000004', '1000027', '1000035', '1000038', '1000051', '100006',
       '100007', '1000072', '1000076',
       ...
       '999919', '999921', '999932', '999935', '999945', '999949', '999972',
       '999977', '999984', '999988'],
      dtype='object', name='Cust_Id', length=341866)

In [ ]:
len(drop_user)

341866

In [ ]:
df_movie_Summary = df_pop.groupby('movie_Id')['Rating'].agg(['count'])
df_movie_Summary

,count
movie_Id,
1,352
2,83
3,1232
4,38
5,831
...,...
4495,325
4496,6276
4497,128


In [ ]:
drop_movie = []

In [ ]:
movie_benchmark = round(df_movie_Summary['count'].quantile(0.7),0)
movie_benchmark

840.0

In [ ]:
drop_movie = df_movie_Summary[df_movie_Summary['count'] < movie_benchmark].index
drop_movie

Index([   1,    2,    4,    5,    6,    7,    9,   10,   11,   12,
       ...
       4483, 4486, 4487, 4489, 4491, 4494, 4495, 4497, 4498, 4499],
      dtype='int64', name='movie_Id', length=3082)

In [ ]:
len(drop_movie)

3082

In [ ]:
df_pop = df_pop[~df_pop['Cust_Id'].isin(drop_user)]
df_pop = df_pop[~df_pop['movie_Id'].isin(drop_movie)]

In [ ]:
df_pop

,Cust_Id,Rating,movie_Id,Title,Genre
693,712664,5.0,3,Grumpier Old Men (1995),Comedy|Romance
694,1331154,4.0,3,Grumpier Old Men (1995),Comedy|Romance
701,2031561,4.0,3,Grumpier Old Men (1995),Comedy|Romance
702,616720,4.0,3,Grumpier Old Men (1995),Comedy|Romance
703,2467008,4.0,3,Grumpier Old Men (1995),Comedy|Romance
...,...,...,...,...,...
23812080,196584,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812082,15737,4.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812083,823628,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812085,1055714,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller


# Model Building

In [ ]:
!pip install scikit-surprise

In [ ]:
from surprise import Reader , Dataset , SVD
from surprise.model_selection import cross_validate
reader = Reader()
data = Dataset.load_from_df(df_pop[['Cust_Id', 'movie_Id', 'Rating']][:100000], reader)
data

In [ ]:
svd = SVD()
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=4)

{'test_rmse': array([0.47669069, 0.47765442, 0.47716096, 0.4761901 ]),
 'test_mae': array([0.45233368, 0.4531893 , 0.4531679 , 0.45605521]),
 'fit_time': (4.413092136383057,
  4.078842401504517,
  1.551156759262085,
  1.5201685428619385),
 'test_time': (0.767545223236084,
  0.4680609703063965,
  0.12704086303710938,
  0.12934136390686035)}

# Recommendation


In [ ]:
df_pop

,Cust_Id,Rating,movie_Id,Title,Genre
693,712664,5.0,3,Grumpier Old Men (1995),Comedy|Romance
694,1331154,4.0,3,Grumpier Old Men (1995),Comedy|Romance
701,2031561,4.0,3,Grumpier Old Men (1995),Comedy|Romance
702,616720,4.0,3,Grumpier Old Men (1995),Comedy|Romance
703,2467008,4.0,3,Grumpier Old Men (1995),Comedy|Romance
...,...,...,...,...,...
23812080,196584,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812082,15737,4.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812083,823628,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller
23812085,1055714,5.0,4496,D.O.A. (1988),Film-Noir|Mystery|Thriller


In [ ]:
# Here we picked up user by customer_ID == 15737 to whom we are going to suggest movies from every genre
user_15737 = df_pop[df_pop['Cust_Id'] == '15737']
user_15737

,Cust_Id,Rating,movie_Id,Title,Genre
92225,15737,4.0,28,Persuasion (1995),Drama|Romance
574042,15737,4.0,171,Jeffrey (1995),Comedy|Drama
667026,15737,5.0,175,Kids (1995),Drama
699316,15737,4.0,187,Party Girl (1995),Comedy
814733,15737,4.0,191,"Scarlet Letter, The (1995)",Drama|Romance
...,...,...,...,...,...
22168733,15737,5.0,4256,"Center of the World, The (2001)",Drama
22759054,15737,4.0,4330,"Scarlet Empress, The (1934)",Drama
23373039,15737,4.0,4418,Things Are Tough All Over (1982),Comedy
23700635,15737,4.0,4472,Bad Dreams (1988),Horror|Thriller


User 15737 have rated 113 movies

In [ ]:
user_15737["Estimated_Score"]= user_15737["movie_Id"].apply(lambda x:svd.predict('15737' ,x).est)
user_15737['Estimated_Score']

/tmp/ipython-input-4183485638.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  user_15737["Estimated_Score"]= user_15737["movie_Id"].apply(lambda x:svd.predict('15737' ,x).est)


,Estimated_Score
92225,4.334725
574042,4.345358
667026,4.345358
699316,4.345358
814733,4.345358
...,...
22168733,4.345358
22759054,4.345358
23373039,4.345358
23700635,4.345358


In [ ]:
user_15737

,Cust_Id,Rating,movie_Id,Title,Genre,Estimated_Score
92225,15737,4.0,28,Persuasion (1995),Drama|Romance,4.334725
574042,15737,4.0,171,Jeffrey (1995),Comedy|Drama,4.345358
667026,15737,5.0,175,Kids (1995),Drama,4.345358
699316,15737,4.0,187,Party Girl (1995),Comedy,4.345358
814733,15737,4.0,191,"Scarlet Letter, The (1995)",Drama|Romance,4.345358
...,...,...,...,...,...,...
22168733,15737,5.0,4256,"Center of the World, The (2001)",Drama,4.345358
22759054,15737,4.0,4330,"Scarlet Empress, The (1934)",Drama,4.345358
23373039,15737,4.0,4418,Things Are Tough All Over (1982),Comedy,4.345358
23700635,15737,4.0,4472,Bad Dreams (1988),Horror|Thriller,4.345358


In [ ]:
# displaying top 5 movies from genre with highest estimated scores
user_15737.sort_values('Estimated_Score', ascending=False).head(5)

,Cust_Id,Rating,movie_Id,Title,Genre,Estimated_Score
10574161,15737,4.0,2112,Grand Canyon (1991),Crime|Drama,4.345358
10784031,15737,4.0,2136,"Nutty Professor, The (1963)",Comedy|Sci-Fi,4.345358
14645782,15737,5.0,2862,Caligula (1979),Drama,4.345358
14397645,15737,5.0,2803,"Pelican Brief, The (1993)",Crime|Drama|Mystery|Romance|Thriller,4.345358
14137162,15737,4.0,2780,"Raven, The (1963)",Comedy|Horror,4.345358


# genres that have received best and worst ratings by user ratings

Best ratings we assume movies that have greater than or equal to 4

In [ ]:
df.loc[(df['Rating'] >= 4),['Genre', 'Title']]


,Genre,Title
1,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
2,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
3,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
6,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
8,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
...,...,...
23813489,Comedy,Dirty Rotten Scoundrels (1988)
23813490,Comedy,Dirty Rotten Scoundrels (1988)
23813493,Comedy,Dirty Rotten Scoundrels (1988)
23813497,Comedy,Dirty Rotten Scoundrels (1988)


Worst Ratings we assuming is less than 3

In [ ]:
df.loc[(df['Rating'] <= 2),['Genre', 'Title']]

,Genre,Title
36,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
39,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
40,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
41,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
50,Adventure|Animation|Children|Comedy|Fantasy,Toy Story (1995)
...,...,...
23813491,Comedy,Dirty Rotten Scoundrels (1988)
23813494,Comedy,Dirty Rotten Scoundrels (1988)
23813496,Comedy,Dirty Rotten Scoundrels (1988)
23813498,Comedy,Dirty Rotten Scoundrels (1988)


## Project Summary

This project focuses on building a movie recommendation system using the Netflix dataset. The primary objective is to identify popular genres, build a collaborative filtering model, and provide personalized movie recommendations.

The key steps involved are:

1.  **Data Loading and Initial Exploration**:
    *   The `combined_data_1.txt` file, containing customer ratings, was loaded into a Pandas DataFrame (`df_User_Id`).
    *   Initial checks were performed for null values and to understand the distribution of ratings.
    *   The number of movies and customers were calculated, and the distribution of ratings (1-5 stars) was analyzed.

2.  **Data Preprocessing and Feature Engineering**:
    *   A `movie_Id` column was extracted from the `Cust_Id` column in `df_User_Id` to link ratings to movies.
    *   Null ratings were removed from `df_User_Id`.
    *   Movie title information was loaded from `movies (1) (1).csv` into `df_titles`, and the header row was dropped. The `movie_Id` column in `df_titles` was converted to a numeric type.
    *   The `df_User_Id` and `df_titles` DataFrames were merged on `movie_Id` to create a comprehensive `df` containing customer ratings, movie titles, and genres.

3.  **Filtering for Popular and Well-Rated Content**:
    *   The DataFrame `df` was filtered to `df_pop` to include only ratings greater than or equal to 4, aiming to focus on highly-rated movies.
    *   Benchmarks were established to filter out users and movies with insufficient activity/ratings:
        *   Users with fewer ratings than the 75th percentile of user ratings (`user_benchmark = 38`) were dropped.
        *   Movies with fewer ratings than the 70th percentile of movie ratings (`movie_benchmark = 840`) were dropped.
    *   This filtering helps to reduce noise and focus on more reliable data for recommendation.

4.  **Model Building (Collaborative Filtering)**:
    *   The `scikit-surprise` library was used to build a collaborative filtering model.
    *   A `Reader` object was initialized, and a `Dataset` was created from the `Cust_Id`, `movie_Id`, and `Rating` columns of the filtered `df_pop` DataFrame.
    *   A Singular Value Decomposition (SVD) model was initialized and evaluated using 4-fold cross-validation with RMSE and MAE metrics.

5.  **Generating Recommendations**:
    *   Recommendations were generated for a specific user (customer ID '15737').
    *   The SVD model predicted an 'Estimated_Score' for movies this user had rated. The top 5 movies with the highest estimated scores for this user were identified.

6.  **Genre Analysis**:
    *   The analysis identified 'Drama' as the most popular and liked genre based on `movie_Id` counts in highly-rated movies.
    *   The notebook also displayed movies and genres that received best (ratings >= 4) and worst (ratings <= 2) ratings, highlighting commonalities and differences in user preferences across these categories.


# Conclusion:
From the above analysis, it is observed that certain movies and genres appear in both the best-rated and worst-rated categories, which indicates that user perception varies depending on the number and diversity of users rating the same content.

The dataset shows that more than 1 crore records have ratings greater than or equal to 4, suggesting that a large volume of users have positively rated many movies. In contrast, approximately 35.16 lakh records have ratings below 3, indicating that comparatively fewer users have given low ratings.

This imbalance highlights that highly rated movies are supported by stronger user participation, whereas poorly rated movies tend to have limited engagement. Additionally, some movies appear exclusively in either the high-rating or low-rating groups, implying that not all content overlaps across these filtered segments and that user preference is not uniformly distributed across all movies and genres.